protocols/protocol/cloudstate/crdt.proto

```diff
@@ -51,20 +56,41 @@ option go_package = "cloudstate/protocol";
 // Special care must be taken when working with maps and sets. The keys/values are google.protobuf.Any, which encodes
 // the value as binary protobuf, however, serialized protobufs are not stable, two semantically equal objects could
 // encode to different bytes. It is the responsibility of the user function to ensure that stable encodings are used.
+//
+//---
+//
+//`Specification`:
+//
+//  - An implementation of `Crdt` is only valid iff:
+//  - `handle`
+//    - TODO: After invoking handle, the first message sent will always be a CrdtInit message, containing the entity ID, and,
+//      if it exists or is available, the current state of the entity. After that, one or more commands may be sent,
+//      as well as deltas as they arrive, and the entire state if either the entity is created, or the proxy wishes the
+//      user function to replace its entire state.
+//    - TODO: The user function must respond with one reply per command in. They do not necessarily have to be sent in the same
+//      order that the commands were sent, the command ID is used to correlate commands to replies.
+//    - TODO
 service Crdt {
-
-    // After invoking handle, the first message sent will always be a CrdtInit message, containing the entity ID, and,
-    // if it exists or is available, the current state of the entity. After that, one or more commands may be sent,
-    // as well as deltas as they arrive, and the entire state if either the entity is created, or the proxy wishes the
-    // user function to replace its entire state.
-    //
-    // The user function must respond with one reply per command in. They do not necessarily have to be sent in the same
-    // order that the commands were sent, the command ID is used to correlate commands to replies.
+    // TODO
     rpc handle(stream CrdtStreamIn) returns (stream CrdtStreamOut);
 }
 
-// Message for the Crdt handle stream in.
+//*CrdtStreamIn* TODO
+//
+//---
+//
+//`Specification`:
+//
+//  - A `CrdtStreamIn` is considered to be valid iff:
+//    - `message` MUST be one of the following:
+//      - `init` TODO
+//      - `state` TODO
+//      - `changed` TODO
+//      - `deleted` TODO
+//      - `command` TODO
+//      - `stream_cancelled` TODO
 message CrdtStreamIn {
+
     oneof message {
 
         // Always sent first, and only once.
```

Comment:
All protocol errors should be reported back to the proxy by sending a Failure followed by closing the stream.

The command id should be left unset, the failure is not in relation to a command, it's in relation to the stream, and since it's sent down the same stream that the protocol error occurred on, it's implicitly correlated.

Also, note that the purpose of correlating errors to commands is so that errors can be sent back to the client that initiated the command. But in the case of a protocol error, the thing that made the error is the proxy, not a client of the proxy. There's nothing to correlate, the proxy just needs to be told you've done the wrong thing, so it can log it, so that a human can manually come, see the log message, debug, and fix the bug in the proxy (or fix the bug in the user function support library that incorrectly identified a protocol error).